src/WorkflowStep.ts

```diff
@@ -0,0 +1,379 @@
+  // Ensure callbackId is valid
+  if (typeof callbackId !== 'string') {
+    const errorMsg = 'WorkflowStep expects a callback_id as the first argument';
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Ensure step config object is passed in
+  if (typeof config !== 'object') {
+    const errorMsg = 'WorkflowStep expects a configuration object as the second argument';
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Check for missing required keys
+  const requiredKeys: (keyof WorkflowStepOptions)[] = ['save', 'edit', 'execute'];
+  const missingKeys: (keyof WorkflowStepOptions)[] = [];
+  requiredKeys.forEach((key) => {
+    if (config[key] === undefined) {
+      missingKeys.push(key);
+    }
+  });
+
+  if (missingKeys.length > 0) {
+    const errorMsg = `WorkflowStep is missing required keys: ${missingKeys.join(', ')}`;
+    throw new WorkflowStepInitializationError(errorMsg);
+  }
+
+  // Ensure a callback or an array of callbacks is present
+  const requiredFns: (keyof WorkflowStepOptions)[] = ['save', 'edit', 'execute'];
+  requiredFns.forEach((fn) => {
+    if (typeof config[fn] !== 'function' && !Array.isArray(config[fn])) {
+      const errorMsg = `WorkflowStep ${fn} property must be a function or an array of functions`;
+      throw new WorkflowStepInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processStepMiddleware()` invokes each callback for lifecycle event
+ * @param args workflow_step_edit action
+ */
+export async function processStepMiddleware(
+  args: AllWorkflowStepMiddlewareArgs,
+  middleware: WorkflowStepMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  // TODO :: revisit type used below (look into contravariance)
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(callbacks, args, context, client, logger, async () =>
+      lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+export function isStepEvent(args: AnyMiddlewareArgs): args is AllWorkflowStepMiddlewareArgs {
+  const validTypes = new Set(['workflow_step_edit', 'workflow_step', 'workflow_step_execute']);
+  return validTypes.has(args.payload.type);
+}
+
+function selectToken(context: Context): string | undefined {
+  return context.botToken !== undefined ? context.botToken : context.userToken;
+}
+
+/**
+ * Factory for `configure()` utility
+ * @param args workflow_step_edit action
+ */
+function createStepConfigure(
+  args: AllWorkflowStepMiddlewareArgs<SlackActionMiddlewareArgs<WorkflowStepEdit>>,
+): StepConfigureFn {
+  const {
+    context,
+    client,
+    body: { callback_id, trigger_id },
+  } = args;
+  const token = selectToken(context);
+
+  return (config: Parameters<StepConfigureFn>[0]) => {
+    const { blocks, private_metadata, submit_disabled, external_id } = config;
```

Comment:
Based on the conversation today, team concluded that not destructuring, and instead sending in all config props sent in by the developer, will allow for immediate support of new features as they arise.